otherlibs/dune-rpc/dune_rpc.mli

```diff
@@ -76,25 +76,60 @@ module V1 : sig
       }
   end
 
-  module Error : sig
-    type t =
-      { target : string option
-      ; message : string
-      ; loc : Loc.t option
-      }
+  module Diagnostic : sig
+    type severity =
+      | Error
+      | Warning
+
+    module Promotion : sig
+      type t =
+        { in_build : string
+        ; in_source : string
+        }
+    end
+
+    type t
+
+    val loc : t -> Loc.t option
+
+    val message : t -> string
+
+    val severity : t -> severity option
+
+    val promotion : t -> Promotion.t list
+
+    val targets : t -> string list
+
+    module Event : sig
+      type nonrec t =
+        | Add of t
+        | Remove of t
```

Comment:
It might be better to associate a diagnostic with a unique (within a session, at least) id and make the message `Remove of Id.t`.

Reply:
Yup, I am thinking of something like that.

> within a session, at least

This is what I'm wondering about. I'm thinking just making the ID the hash of all the fields. I'd prefer if the Id's persist across sessions as otherwise it's a headache for clients to always make sure they don't store stale Id's.

>  id and make the message Remove of Id.t.

What about just making this transparent? The user passes the error, but the protocol passes the ID under the hood. Id's sometimes have tricky semantics, and I hate letting the user know of them.

Reply:
>  I'd prefer if the Id's persist across sessions as otherwise it's a headache for clients to always make sure they don't store stale Id's.

This makes a lot of sense :+1:

> The user passes the error, but the protocol passes the ID under the hood.

I'm of two minds about this. Normally, I am certainly all for keeping this kind of thing transparent for exactly the reason you suggest. On the other hand, without the id, if a client is maintaining a set of diagnostics, handling a `Remove t` event would require inspecting all of the error fields.